edlib/src/edlib.tpp

```diff
@@ -201,9 +226,26 @@ EdlibAlignResult edlibAlign(const Element* const queryOriginal, const int queryL
     /*--------------------- INITIALIZATION ------------------*/
     int maxNumBlocks = ceilDiv(queryLength, WORD_SIZE); // bmax in Myers
     int W = maxNumBlocks * WORD_SIZE - queryLength; // number of redundant cells in last level blocks
-    // TODO: According to the future implementation of EqualityDefinition we need to modify instantiation
-    //EqualityDefinition equalityDefinition(elementToAlphabetIdx, config.additionalEqualities, config.additionalEqualitiesLength);
-    EqualityDefinition<AlphabetIdx> equalityDefinition;
+    // The boolean array, hasAdditionalEqualities shows which alphabet indexes has at least one
+    // additional equality. It accelerates checking equality in 'areEqual' function.
+    bool* hasAdditionalEquality =  static_cast<bool *>(malloc(sizeof(bool) * elementToAlphabetIdx.size()));
+    for (AlphabetIdx j = 0; j < elementToAlphabetIdx.size(); j++){
+        hasAdditionalEquality[j] = false; // initialize the boolean array
+    }
+    if (config.additionalEqualities != NULL) {
+        for (AlphabetIdx i = 0; i < config.additionalEqualitiesLength; i++) {
+            AlphabetIdx a = elementToAlphabetIdx[config.additionalEqualities[i].first];
+            AlphabetIdx b = elementToAlphabetIdx[config.additionalEqualities[i].second];
+            // a and b have at least one additional equality
+            hasAdditionalEquality[a] = true;
+            hasAdditionalEquality[b] = true;
+        }
+    }
+    SetOfPairs<AlphabetIdx> equalitySet =
+            makeEqualitySet<Element, AlphabetIdx>(elementToAlphabetIdx,
+                                                  config.additionalEqualities,
+                                                  config.additionalEqualitiesLength);
+    EqualityDefinition<AlphabetIdx> equalityDefinition(equalitySet, hasAdditionalEquality);
```

Comment:
Looking at all this, I think the approach / algorithm is good, however it feels (as mentioned in some comments above) that we could encapsulate some things better, so let's see how to do that.

We have config.additionalEqualities array, we have its size, and we have elementToAlphabetIdx map. If I am correct, that is all that we need to construct the EqualityDefinition object, that will then allow us to call .areEqual(a, b) to check if two alphabet indexes are equal, right?
If that is so, then in edlibAlign function, we would ideally have only the following logic: passing inputs (config.additionalEqualities + their size, elementToAlphabetIdx, maybe smth else I missed) to a function, and receiving as a result EqualityDefinition object! Nothing more, no internals exposed. Then, we use that object where needed.

Now, the function that generates EqualityDefinition -> it could either be constructor of EqualityDefinition, in which case most of the logic will go into constructor, or it could be standalone function and constructor of EqualityDefinition will be as simple as it is now. I personally prefer doing stuff in constructor, but I also heard opposite opinions so I am not 100% sure at the moment what is better approach, probably it is not very important.

Also, in this approach, important thing is that EqualityDefinition would worry about allocating and deallocating memory it needs, reducing chance of memory leaks.

Reply:
@Martinsos Let me explain why I put the function that generates `EqualityDefinition` out of its constructor. As you mentioned the `EqualityDefinition` has just one method and that is `areEqual`. When we call `.areEqual` we have to pass two variables of type `AlphabetIdx` to be compared so knowing the custom type of `AlphabetIdx` is enough and there is no need to carry `Element`. `areEqual` has nothing to do with `Element`. 
On the other hand, the pairs in `additionalEqualities` are of type `Element` and generating `EqualityDefintion` needs mapping those pairs from the type `Element` to `AlphabetIdx`. If we put this mapping procedure in the constructor we have to add a template argument to `EqualityDefinition` for `Element`, which is needed just for generating the object.
I decided to avoid adding this template argument. Maybe you have better idea? (or you think we can have it?)

Reply:
Based on what you said I put all the procedure for generating `EqualityDefintion` in a single function. 
This is the new commit -> https://github.com/Martinsos/edlib/pull/154/commits/4e6c4dd5595d8aca07bab7408d84afa014b715a5

Reply:
Aha, good point with the template argument, I missed that!
Makes sense then, I agree with the approach of external constructor. Although it maybe somewhat complicates the situation with allocating the memory hm, since I think now we might have some copying happening with vectors/sets, although I am not 100% sure. While if it was happening in constructor, it would be easier, because vector/set would get allocated as object is constructed and you would just fill them and that is it.

Hm here is a thought: if we first transformed additionalEqualities from EdlibEqualityPair<Element>* into vector<EdlibEqualityPair<AlphabetIdx>>, which is very easy to do, we could just pass that one vector to constructor of EqualityDefinition and that is it! It would take that vector by reference, so no copying, and then it would just fill the set and vector it already has as private members.

Hm this sounds great to me, what do you think? So to summarize:
1. Function that takes EdlibEqualityPair<Element>* and its length and elementToAlphabetIdx and returns vector<EdlibEqualityPair<AlphabetIdx>>. Or maybe we can even inline this code, whatever you like better.
2. Constructor in EqualityDefinition that takes vector<EdlibEqualityPair<AlphabetIdx>> by reference and constructs equalitySet and hasAdditionalEqualities.

Reply:
All agreed!